service/evidence/evidence_store.go

```diff
@@ -43,60 +46,109 @@ var log *logrus.Entry
 
 // Service is an implementation of the Clouditor req service (evidenceServer)
 type Service struct {
-	// Currently only in-memory
-	evidences map[string]*evidence.Evidence
+	storage persistence.Storage
 
 	// evidenceHooks is a list of hook functions that can be used if one wants to be
 	// informed about each evidence
 	evidenceHooks []evidence.EvidenceHookFunc
 	// mu is used for (un)locking result hook calls
 	mu sync.Mutex
 
+	// authz defines our authorization strategy, e.g., which user can access which cloud service and associated
+	// resources, such as evidences and assessment results.
+	authz service.AuthorizationStrategy
+
 	evidence.UnimplementedEvidenceStoreServer
 }
 
-func NewService() *Service {
-	return &Service{
-		evidences: make(map[string]*evidence.Evidence),
+func WithStorage(storage persistence.Storage) service.Option[Service] {
+	return func(svc *Service) {
+		svc.storage = storage
+	}
+}
+
+func NewService(opts ...service.Option[Service]) (svc *Service) {
+	var (
+		err error
+	)
+	svc = new(Service)
+
+	for _, o := range opts {
+		o(svc)
+	}
+
+	// Default to an allow-all authorization strategy
+	if svc.authz == nil {
+		svc.authz = &service.AuthorizationStrategyAllowAll{}
+	}
+
+	if svc.storage == nil {
+		svc.storage, err = inmemory.NewStorage()
+		if err != nil {
+			log.Errorf("Could not initialize the storage: %v", err)
+		}
 	}
+
+	return
 }
 
 func init() {
 	log = logrus.WithField("component", "Evidence Store")
 }
 
 // StoreEvidence is a method implementation of the evidenceServer interface: It receives a req and stores it
-func (s *Service) StoreEvidence(_ context.Context, req *evidence.StoreEvidenceRequest) (resp *evidence.StoreEvidenceResponse, err error) {
+func (svc *Service) StoreEvidence(ctx context.Context, req *evidence.StoreEvidenceRequest) (res *evidence.StoreEvidenceResponse, err error) {
 	// Validate request
 	err = service.ValidateRequest(req)
 	if err != nil {
-		newError := fmt.Errorf("invalid evidence: %w", err)
-		log.Error(newError)
+		err = fmt.Errorf("invalid evidence: %w", err)
+		log.Error(err)
 
-		go s.informHooks(nil, newError)
+		go svc.informHooks(nil, err)
 
-		resp = &evidence.StoreEvidenceResponse{
+		res = &evidence.StoreEvidenceResponse{
 			Status:        false,
-			StatusMessage: newError.Error(),
+			StatusMessage: err.Error(),
 		}
 
-		return resp, status.Errorf(codes.InvalidArgument, "%v", newError)
+		err = status.Errorf(codes.InvalidArgument, "%v", err)
+		return
+	}
+
+	// Check, if this request has access to the cloud service according to our authorization strategy.
+	if !svc.authz.CheckAccess(ctx, service.AccessUpdate, req) {
+		return nil, service.ErrPermissionDenied
 	}
 
-	s.evidences[req.Evidence.Id] = req.Evidence
-	go s.informHooks(req.Evidence, nil)
+	err = svc.storage.Create(req.Evidence)
+	if err != nil {
+		// TODO(lebogg): Almost identical error handling as above -> extract here or even to internal
```

Comment:
This code here by @lebogg might need to be adjusted to the "new" behaviour of our streaming functions such as AssessEvidence.